// @ontologies/foaf/index.js
import { createNS } from "@ontologies/core";
/** Function to create arbitrary terms within the 'Friend of a Friend (FOAF) vocabulary' ontology */
export const ns = createNS("http://xmlns.com/foaf/0.1/");
/* Classes */
/** A foaf:LabelProperty is any RDF property with texual values that serve as labels. */
export const LabelProperty = /*#__PURE__*/ ns("LabelProperty");
/** A person. */
export const Person = /*#__PURE__*/ ns("Person");
/** An agent (eg. person, group, software or physical artifact). */
export const Agent = /*#__PURE__*/ ns("Agent");
/** A document. */
export const Document = /*#__PURE__*/ ns("Document");
/** An organization. */
export const Organization = /*#__PURE__*/ ns("Organization");
/** A class of Agents. */
export const Group = /*#__PURE__*/ ns("Group");
/** A project (a collective endeavour of some kind). */
export const Project = /*#__PURE__*/ ns("Project");
/** An image. */
export const Image = /*#__PURE__*/ ns("Image");
/** A personal profile RDF document. */
export const PersonalProfileDocument = /*#__PURE__*/ ns("PersonalProfileDocument");
/** An online account. */
export const OnlineAccount = /*#__PURE__*/ ns("OnlineAccount");
/** An online gaming account. */
export const OnlineGamingAccount = /*#__PURE__*/ ns("OnlineGamingAccount");
/** An online e-commerce account. */
export const OnlineEcommerceAccount = /*#__PURE__*/ ns("OnlineEcommerceAccount");
/** An online chat account. */
export const OnlineChatAccount = /*#__PURE__*/ ns("OnlineChatAccount");
/* Properties */
/** A
personal mailbox, ie. an Internet mailbox associated with exactly one owner, the first owner of this mailbox. This is a 'static inverse functional property', in that  there is (across time and change) at most one individual that ever has any particular value for foaf:mbox. */
export const mbox = /*#__PURE__*/ ns("mbox");
/** The sha1sum of the URI of an Internet mailbox associated with exactly one owner, the  first owner of the mailbox. */
export const mbox_sha1sum = /*#__PURE__*/ ns("mbox_sha1sum");
/** The gender of this Agent (typically but not necessarily 'male' or 'female'). */
export const gender = /*#__PURE__*/ ns("gender");
/** A textual geekcode for this person, see http://www.geekcode.com/geek.html */
export const geekcode = /*#__PURE__*/ ns("geekcode");
/** A checksum for the DNA of some thing. Joke. */
export const dnaChecksum = /*#__PURE__*/ ns("dnaChecksum");
/** A sha1sum hash, in hex. */
export const sha1 = /*#__PURE__*/ ns("sha1");
/** A location that something is based near, for some broadly human notion of near. */
export const based_near = /*#__PURE__*/ ns("based_near");
/** Title (Mr, Mrs, Ms, Dr. etc) */
export const title = /*#__PURE__*/ ns("title");
/** A short informal nickname characterising an agent (includes login identifiers, IRC and other chat nicknames). */
export const nick = /*#__PURE__*/ ns("nick");
/** A jabber ID for something. */
export const jabberID = /*#__PURE__*/ ns("jabberID");
/** An AIM chat ID */
export const aimChatID = /*#__PURE__*/ ns("aimChatID");
/** A Skype ID */
export const skypeID = /*#__PURE__*/ ns("skypeID");
/** An ICQ chat ID */
export const icqChatID = /*#__PURE__*/ ns("icqChatID");
/** A Yahoo chat ID */
export const yahooChatID = /*#__PURE__*/ ns("yahooChatID");
/** An MSN chat ID */
export const msnChatID = /*#__PURE__*/ ns("msnChatID");
/** A name for some thing. */
export const name = /*#__PURE__*/ ns("name");
/** The first name of a person. */
export const firstName = /*#__PURE__*/ ns("firstName");
/** The last name of a person. */
export const lastName = /*#__PURE__*/ ns("lastName");
/** The given name of some person. */
export const givenName = /*#__PURE__*/ ns("givenName");
/** The given name of some person. */
export const givenname = /*#__PURE__*/ ns("givenname");
/** The surname of some person. */
export const surname = /*#__PURE__*/ ns("surname");
/** The family name of some person. */
export const family_name = /*#__PURE__*/ ns("family_name");
/** The family name of some person. */
export const familyName = /*#__PURE__*/ ns("familyName");
/** A phone,  specified using fully qualified tel: URI scheme (refs: http://www.w3.org/Addressing/schemes.html#tel). */
export const phone = /*#__PURE__*/ ns("phone");
/** A homepage for some thing. */
export const homepage = /*#__PURE__*/ ns("homepage");
/** A weblog of some thing (whether person, group, company etc.). */
export const weblog = /*#__PURE__*/ ns("weblog");
/** An OpenID for an Agent. */
export const openid = /*#__PURE__*/ ns("openid");
/** A tipjar document for this agent, describing means for payment and reward. */
export const tipjar = /*#__PURE__*/ ns("tipjar");
/** A .plan comment, in the tradition of finger and '.plan' files. */
export const plan = /*#__PURE__*/ ns("plan");
/** Something that was made by this agent. */
export const made = /*#__PURE__*/ ns("made");
/** An agent that
made this thing. */
export const maker = /*#__PURE__*/ ns("maker");
/** An image that can be used to represent some thing (ie. those depictions which are particularly representative of something, eg. one's photo on a homepage). */
export const img = /*#__PURE__*/ ns("img");
/** A depiction of some thing. */
export const depiction = /*#__PURE__*/ ns("depiction");
/** A thing depicted in this representation. */
export const depicts = /*#__PURE__*/ ns("depicts");
/** A derived thumbnail image. */
export const thumbnail = /*#__PURE__*/ ns("thumbnail");
/** A Myers Briggs (MBTI) personality classification. */
export const myersBriggs = /*#__PURE__*/ ns("myersBriggs");
/** A workplace homepage of some person; the homepage of an organization they work for. */
export const workplaceHomepage = /*#__PURE__*/ ns("workplaceHomepage");
/** A work info homepage of some person; a page about their work for some organization. */
export const workInfoHomepage = /*#__PURE__*/ ns("workInfoHomepage");
/** A homepage of a school attended by the person. */
export const schoolHomepage = /*#__PURE__*/ ns("schoolHomepage");
/** A person known by this person (indicating some level of reciprocated interaction between the parties). */
export const knows = /*#__PURE__*/ ns("knows");
/** A page about a topic of interest to this person. */
export const interest = /*#__PURE__*/ ns("interest");
/** A thing of interest to this person. */
export const topic_interest = /*#__PURE__*/ ns("topic_interest");
/** A link to the publications of this person. */
export const publications = /*#__PURE__*/ ns("publications");
/** A current project this person works on. */
export const currentProject = /*#__PURE__*/ ns("currentProject");
/** A project this person has previously worked on. */
export const pastProject = /*#__PURE__*/ ns("pastProject");
/** An organization funding a project or person. */
export const fundedBy = /*#__PURE__*/ ns("fundedBy");
/** A logo representing some thing. */
export const logo = /*#__PURE__*/ ns("logo");
/** A topic of some page or document. */
export const topic = /*#__PURE__*/ ns("topic");
/** The primary topic of some page or document. */
export const primaryTopic = /*#__PURE__*/ ns("primaryTopic");
/** The underlying or 'focal' entity associated with some SKOS-described concept. */
export const focus = /*#__PURE__*/ ns("focus");
/** A document that this thing is the primary topic of. */
export const isPrimaryTopicOf = /*#__PURE__*/ ns("isPrimaryTopicOf");
/** A page or document about this thing. */
export const page = /*#__PURE__*/ ns("page");
/** A theme. */
export const theme = /*#__PURE__*/ ns("theme");
/** Indicates an account held by this agent. */
export const account = /*#__PURE__*/ ns("account");
/** Indicates an account held by this agent. */
export const holdsAccount = /*#__PURE__*/ ns("holdsAccount");
/** Indicates a homepage of the service provide for this online account. */
export const accountServiceHomepage = /*#__PURE__*/ ns("accountServiceHomepage");
/** Indicates the name (identifier) associated with this online account. */
export const accountName = /*#__PURE__*/ ns("accountName");
/** Indicates a member of a Group */
export const member = /*#__PURE__*/ ns("member");
/** Indicates the class of individuals that are a member of a Group */
export const membershipClass = /*#__PURE__*/ ns("membershipClass");
/** The birthday of this Agent, represented in mm-dd string form, eg. '12-31'. */
export const birthday = /*#__PURE__*/ ns("birthday");
/** The age in years of some agent. */
export const age = /*#__PURE__*/ ns("age");
/** A string expressing what the user is happy for the general public (normally) to know about their current activity. */
export const status = /*#__PURE__*/ ns("status");
/* Other terms */
