// @inrupt/solid-client-authn-core/dist/login/oidc/refresh/__mocks__/TokenRefresher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mockDefaultTokenRefresher = exports.mockDefaultTokenSet = exports.mockTokenRefresher = void 0;
const globals_1 = require("@jest/globals");
const mockTokenRefresher = (tokenSet) => {
    return {
        refresh: globals_1.jest
            .fn()
            .mockResolvedValue(tokenSet),
    };
};
exports.mockTokenRefresher = mockTokenRefresher;
const mockIdTokenPayload = () => {
    return {
        sub: "https://my.webid",
        iss: "https://my.idp/",
        aud: "https://resource.example.org",
        exp: 1662266216,
        iat: 1462266216,
    };
};
const mockDefaultTokenSet = () => {
    return {
        accessToken: "some refreshed access token",
        idToken: JSON.stringify(mockIdTokenPayload()),
    };
};
exports.mockDefaultTokenSet = mockDefaultTokenSet;
const mockDefaultTokenRefresher = () => (0, exports.mockTokenRefresher)((0, exports.mockDefaultTokenSet)());
exports.mockDefaultTokenRefresher = mockDefaultTokenRefresher;
